<a href="https://colab.research.google.com/github/carlosdelafuente/03MIAR---Algoritmos-de-Optimizacion---2021-/blob/main/Seminario/Seminario_Carlos_de_la_Fuente_Torres.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos: Carlos de la Fuente Torres   <br>
Url: https://github.com/.../03MAIR---Algoritmos-de-Optimizacion---2019/tree/master/SEMINARIO<br>
Problema:
> **3. Combinar cifras y operaciones**

Descripción del problema:

• El problema consiste en analizar el siguiente problema y diseñar un algoritmo que lo resuelva.

• Disponemos de las 9 cifras del 1 al 9 (excluimos el cero) y de los 4 signos básicos de las
operaciones fundamentales: suma(+), resta(-), multiplicación(*) y división(/).

• Debemos combinarlos alternativamente sin repetir ninguno de ellos para obtener una cantidad
dada. Un ejemplo sería para obtener el 4:

>**4+2-6/3*1 = 4**





                                        

##(*)¿Cuántas posibilidades hay sin tener en cuenta las restricciones?<br>



##¿Cuántas posibilidades hay teniendo en cuenta todas las restricciones.




#### Posibilidades sin tener en cuenta las restricciones:

La restricción que tenemos es **no repetir ningún número y ningún signo**, por ello, si eliminamos todas las restricciones podemos repetir **TODOS** los números y **TODOS** los signos:
> Esto nos deja con 9^5 posibilidades para los números y 4^4 posibilidades para los signos, lo que nos hace un total de **(9^5)*(4^4) = 15116544 posibilidades.**


#### Posibilidades teniendo en cuenta las restricciones:

Si tenemos en cuenta las restricciones anteriores (no repetir ni números ni signos) nos encontramos con el siguiente espacio de posibilidades:

Al no poder repetir números y signos la **condición** principal de nuestra solución es que una vez elegido un número o un signo en la siguiente elección tenemos una posibilidad menos. ¿En qué se traduce esto?:

> En cuento a los **números**, si partimos de 9 posibilidades en la primera posición y de una posibilidad menos cada vez que avancemos,sabiendo que solamente se cogen 5 números, las posibilidades son 9! quitando las 4 últimas posibilidades, lo que  matemáticamente se traduce en: **9!/4!**

> En el caso de los **signos**, al tener que 4 signos estamos ante **4!** posibilidades.

Por tanto, si unimos ambas probabilidades, nos encontramos con un total de (9!/4!)*4! = 9! posibilidades.

**9! = 362880 posibilidades**

## Modelo para el espacio de soluciones<br>


> 

### (*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)

Aunque a priori planteé el problema con una única lista como estructura de datos para almacenar tanto número como signos me dí cuenta de que era poco manejable al tratar las restricciones de números y signos a la vez. Por ello, tras plantearlo mejor, **me decante por almacenar los números en una lista y los signos en otra y a su vez ordenados de menor a mayor**. De este modo el control sobre las variables es mucho más sencillo y mucho más fácil de programar.

Sabiendo que se deben intercalar los signos (en posiciones pares) y los números (en posiciones impares) el problema se simplifica mucho para cualquier tipo de planteamiento ya que eliminamos comprobaciones mientras construimos la soluciión y solo debemos chequear la solución una vez.

Ordenarlos de menor a mayor supone una ventaja principal en cuento al planteamiento. Si tenemos los números y signos ordenados de menor a mayor, siendo el 1 y '/' los menores y el 9 y '*' los mayores, tenemos una estructura de datos definida que en el caso de un algoritmo de fuerza bruta no es una implementación crucial pero si en el caso de otras implemetanciones.
> 1) Algoritmos de búsqueda: Para algoritmos BFS, DFS o de ramificación y poda para poder ir siempre del menor al mayor o definir bien la cota inferior o superior.

> 2) Algoritmos Heurísticos/estocásticos: Para ponderar con mayor o menor probabilidad los elementos de la lista dependiendo de las reglas establacidas. (Por ejemplo, dependiendo de lo lejos o cerca que este la solución actual de la solución óptima).

In [ ]:
numeros = ['1','2','3','4','5','6','7','8','9']
signos  = ['/', '-', '+', '*']

##Según el modelo para el espacio de soluciones<br>


###(*)¿Cual es la función objetivo?

En este problema en concreto buscamos una solución específica, por tanto, **no tenemos una función objetivo** ni la posibilidade de cuantificar si nuestra solución se aproxima o no a la solución óptima. En este problema en concreto se debe encontrar la solución al problema. El cual puede tener o no solución con las restricciones que tenemos y podemos llegar o no a la solución dependiendo del algoritmo que se utilice.

###(*)¿Es un problema de maximización o minimización?

No se trata **ni de un problema de maximización ni de un problema de minimización**. Para llegar a la solución no debemos ni maximizar ni minimizar una función objetivo ya que tenemos una solución dada. Debemos de ser capaces de dar o no dar la solución. O hay solución o no hay solución.

##Diseña un algoritmo para resolver el problema por fuerza bruta

Al tratarse de un algoritmo de fuerza bruta vamos a explorar todas las soluciones posibles dadas las restricciones del problema:

Para ello creamos una lista con todas las posibles combinaciones de números y tolas las posibles combinaciones de signos. Donde debemos recordar que en el caso de números tenemos un total de 9!/4! = 15120 posibilidades y en el caso de los signos tenemos un total de 4! = 24 posibilidades:

In [ ]:
from itertools import permutations

# creamos una lista con todas las posibles combinaciones de números, recordando que
combinaciones_numeros = list(permutations(numeros, 5))
print("Total combinaciones números: ",len(combinaciones_numeros))
combinaciones_signos = list(permutations(signos, 4))
print("Total combinaciones signos : ",len(combinaciones_signos))

Total combinaciones números:  15120
Total combinaciones signos :  24


In [ ]:
def combinar_numeros_signos(nums, sigs):
  expression = ''
  for n, s in zip(range(len(nums)-1), range(len(sigs))):
    expression = expression + nums[n]
    expression = expression + sigs[s]
  expression = expression + nums[-1]
  return expression

In [ ]:
def find_expression_fuerza_bruta(comb_num, comb_sig, value):
  for current_nums in comb_num:
    for current_sigs in comb_sig:

      expression = combinar_numeros_signos(current_nums, current_sigs)
      if eval(expression) == value: return expression
      
  return None

In [ ]:
value = 4
expression = find_expression_fuerza_bruta(combinaciones_numeros, combinaciones_signos, value)
if expression is not None:
  print("La expresión", expression, "obtiene como resultado en número", value, " :)")
else:
  print("No hemos podido encontrar una expresión que obtenga el número", value, " :(")

La expresión 1-2*3/6+4 obtiene como resultado en número 4  :)


###Calcula la complejidad del algoritmo por fuerza bruta

Al tratarse de un problema con un espacio de posibilidades finito, exactamente 9! posibilidades, somos capaces de cuantificar el coste = **O(n!)**.

La explicación de esto la hemos dado antes:
Para cada valor elegido de nuestra expresión final debemos explorar n posibilidades menos una (la seleccionada en la actual iteración) y así sucesivamente hasta llegar al último valor de la expresión. A nivel matemático esto se expresa como n! permutaciones posibles.

## ¿Qué valor máximo y mínimo se pueden obtener según las condiciones del problema?

Dado que con fuerza bruta podemos explorar todo el espacio de soluciones utilizaremos este algoritmo para calcular el valor más alto y más bajo posible dadas las restricciones y poder responder a las preguntas planteadas:

In [ ]:
import sys

def max_min_fuerza_bruta(comb_num, comb_sig):
  max = min = 0
  exp_min = exp_max = ''

  for current_nums in comb_num:
    for current_sigs in comb_sig:

      expression = combinar_numeros_signos(current_nums, current_sigs)
      if eval(expression) < min:
        min = eval(expression)
        exp_min = expression
      if eval(expression) > max:
        max = eval(expression)
        exp_max = expression

  return min, max, exp_min, exp_max

In [ ]:
min,max,exp_min,exp_max = max_min_fuerza_bruta(combinaciones_numeros, combinaciones_signos)
print("Con", exp_min, "obtenemos el valor mínimo posbile que es", min)
print("Con", exp_max, "obtenemos el valor máximo posbile que es", max)

Con 1+2/7-8*9 obtenemos el valor mínimo posbile que es -70.71428571428571
Con 7-1/6+8*9 obtenemos el valor máximo posbile que es 78.83333333333333


El valor **máximo** que se puede obtener es el **78.83** y el valor **mínimo** que se puede obener es el **-70.71**

 ## ¿Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo ?

La mejor forma de responder a eso es encontrando un contraejemplo:

In [ ]:
def fuerza_bruta_entero(comb_num, comb_sig, value):
  for current_nums in comb_num:
    for current_sigs in comb_sig:

      expression = combinar_numeros_signos(current_nums, current_sigs)
      if int(eval(expression)) == value: return expression
      
  return None

In [ ]:
count = 0
for i in range(int(min),int(max)):
  expression = fuerza_bruta_entero(combinaciones_numeros, combinaciones_signos, 
                                            i)
  if expression is None:
    print("Hemos encontrado el contraejemplo con el valor:", i)
    break
  else:
    count += 1
else:
  print("Podemos encontrar los", count, "valores entre el mínimo y el máximo!! :)")

Podemos encontrar los 148 valores entre el mínimo y el máximo!! :)


**SI** es posible encontrar todos los valores enteros entre el mínimo y el máximo

## (*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

>Antes de echarles un vistazo a los apuntes plantee el problema como un problema de **búsqueda** donde mi primer pensamiento fue utilizar el algoritmo de **ramificación y poda** pero rápidamente llegue a lo conclusión de que quizás no es un buen planteamiento ya que la principal función de este algoritmo es **maximizar** o **minimizar** una función y este NO es un problema ni de maximización ni de minimización. Al igual que, aunque definamos una cota superior y una cota inferior no somos capaces de saber, a priori, si entre esas dos cotas vamos a tener todo espacio de posibilidades y por tanto no somos capaces de saber si la solución óptima se encuentra entre dichas cotas lo que implicaría tener que recorrer todo el árbol hasta las hojas y la complejidad del algoritmo sería **aproximadamente igual** a la de **fuerza bruta**.

>Por el mismo motivo descarté utilizar **backtracking** ya que deberíamos recorrer el árbol hasta las hojas para conocer realmente si hemos o no encontrado la solución óptima.

>Posibles solución: **Búsqueda aleatoria** o **algoritmos genéticos**. En este caso, por simplicidad, y sabiendo cuantificar el coste del algoritmo, utilizaremos como solución búsqueda aleatoria: En este caso, el algoritmo es muy sencillo y se basa en un único paso:

>>1) Seleccionar una combinación de números y una combinación de signos aleatoriamente.



In [ ]:
import random

def busqueda_aleatoria(nums, sigs, value, iters):
  for i in range(iters):
    random_nums = random.choice(nums)
    random_sigs = random.choice(sigs)

    expression = combinar_numeros_signos(random_nums, random_sigs)
    if eval(expression) == value: return expression
      
  return None

In [ ]:
value = 4
expression = busqueda_aleatoria(combinaciones_numeros, combinaciones_signos, value, 300)
if expression is not None:
  print("La expresión", expression, "obtiene como resultado en número", value, " :)")
else:
  print("No hemos podido encontrar una expresión que obtenga el número", value, " :(")

La expresión 1+3*8/2-9 obtiene como resultado en número 4  :)


###(*)Calcula la complejidad del algoritmo 

Como hemos anotado antes, la complejidad de un algoritmo de búsqueda aleatoria es fácilmente cuantificable ya que dependemos del número de iteraciones que realizamos. Si anotamos el número de operaciones como **n** la complejidad del problema es lineal ya que dependemos de un único bucle y por tanto matemáticamente  la complejidad es **O(n)** donde en el mejor de los casos (una única iteración) la complejidad sería de **O(1)** aunque la probabilidad de encontrar la solución sería muy muy baja.

## Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios

Si nos ceñimos a las **restricciones** dadas por el problema podemos variar de entrada únicamente  el **valor** que queremos buscar y el número de **iteraciones**. Una posible implementación es la siguiente:

In [ ]:
dato_aleatorio = random.randint(int(min),int(max))
iteraciones_aleatorias = random.randint(1,1000)

print("Valor aleatorio: ", dato_aleatorio)
print("Iteraciones aleatorias: ", iteraciones_aleatorias)

Valor aleatorio:  41
Iteraciones aleatorias:  388


Por otro lado, si **NO** nos ceñimos a las **restricciones** dadas por el problema podemos variar de entrada el **valor**, las **iteraciones**, la **lista de números posibles** e incluso la **lista de signos**. ¿Cómo? Podríamos tener una lista **n** de números aleatorios y un lista **m** de signos aleatorios. Una posible implementación sería la siguiente:

In [ ]:
total_posibles_signos = ['.', '/', '-', '+', '*']
singos_aleatorios = random.sample(total_posibles_signos,
                                  random.randint(3, len(total_posibles_signos)-1))

total_posibles_numeros = ['1','2','3','4','5','6','7','8','9','10']
numeros_aleatorios = random.sample(total_posibles_numeros,
                                  random.randint(len(total_posibles_signos)+1, 
                                                 len(total_posibles_numeros)))
comb_signos_aleatorios = list(permutations(singos_aleatorios))
comb_numeros_aleatorios = list(permutations(numeros_aleatorios, len(singos_aleatorios)+1))

min_al,max_al,_,_ = max_min_fuerza_bruta(comb_numeros_aleatorios, 
                                       comb_signos_aleatorios)
dato_aleatorio = random.randint(int(min_al),int(max_al))
iteraciones_aleatorias = random.randint(100,10000)

print("Valor aleatorio: ", dato_aleatorio)
print("Iteraciones aleatorias: ", iteraciones_aleatorias)
print("Lista de signos: ", singos_aleatorios)
print("Lista de números: ", numeros_aleatorios)

Valor aleatorio:  -10
Iteraciones aleatorias:  5530
Lista de signos:  ['-', '*', '/', '.']
Lista de números:  ['6', '10', '8', '7', '4', '5', '1', '2', '9']


Si nos fijamos, nuestros nuevos conjuntos de datos añaden una nueva posibilidad en cuento a los signos, el "." que nos permite introducir decimales en la ecuación, y una nueva posibilidad en los números, el 10. Se generan aleatoriamente las listas de signos y números y se define el máximo y el mínimo posible para poder escoger (aleatoriamente) un número entre ellos.
**NOTA:** A diferencia de la ecuación con restricciones, con el nuevo conjunto de datos generados aleatoriamente no podemos garantizar que para todos los números comprendidos entre el máximo y el mínimo haya solución.

## Aplica el algoritmo al juego de datos generado

### **CON restricciones**:

In [ ]:
expression = busqueda_aleatoria(combinaciones_numeros, combinaciones_signos, 
                                dato_aleatorio, iteraciones_aleatorias)
if expression is not None:
  print("La expresión", expression, "obtiene como resultado", dato_aleatorio, " :)")
else:
  print("No hemos podido encontrar una expresión que obtenga", dato_aleatorio, " :(")

La expresión 4/2+9-3*7 obtiene como resultado -10  :)


### **SIN restricciones**:

In [ ]:
expression = busqueda_aleatoria(comb_numeros_aleatorios, comb_signos_aleatorios, 
                                dato_aleatorio, iteraciones_aleatorias)
if expression is not None:
  print("La expresión", expression, "obtiene como resultado", dato_aleatorio, " :)")
else:
  print("No hemos podido encontrar una expresión que obtenga", dato_aleatorio, " :(")

La expresión 1*4/8-10.5 obtiene como resultado -10  :)


## Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo

Toda la implementación del código se ha realizado a partir de lo impartido en clases y a partir de los apuntes de la asignatura.

## Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

Si bien es cierto que el motivo principal de elegir una solución con búsqueda aleatoria ha sido por la simplicidad del problema. Otra solución más elaborada, a partir de una búsqueda aleatoria podría ser utilizar **algoritmos genéticos** sabiendo lo cerca o lejos que esta nuestra solución parcial de la solución óptima. Para ello podríamos jugar con el valor de los números y la magnitud de los signos. Para ello podríamos definir las siguientes reglas:

> **1) Mutación de aproximación:** Si el individuo actual está relativamente a mucha distancia del óptimo el cambio que más positivamente puede afectar es mutar con mayor probabilidad el signo '*' o '/' a las últimas posiciones lo que provocaría cambios más bruscos en el individuo. En caso contrario, si el individuo está próximo a la solución, un movimiento acertado sería mutar los signos '+' y '-'. Idénticamente podríamos aplicar esta metodología a los números.

> **2)Agregación de refinamiento:** Finalmente, si no se ha llegado a la solución óptima, podríamos intercambiar un número de la ecuación por otro muy próximo (teniendo en cuenta la restricción de que no puede ser una que ya se encuentra dentro de la expresión).

La idea principal de avanzar por soluciones heurísticas se centra en el coste del algoritmo ya que para cualquier tamaño de números y signos seguirá siendo un coste lineal de **O(n)**.